Extend Gemini to include getting file content and logging out of GOA

davner opened the pull request:
This PR extends the `Gemini` module. 
1. `get_file_content` - Grabs the content of a file, does not save to disk.
2. `logout` - Deletes the Gemini Observatory Archive cookie.
3. `get_file_url` - helper method to get the URL for direct file download.

I have updated the documentation to include the new functions and included tests.

davner committed:
Extend Gemini to include getting file content and logging out of the Archive.

pep8speaks commented:
Hello @davner! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`astroquery/gemini/tests/test_gemini.py`](https://github.com/astropy/astroquery/blob/ade3a79574247c35b2f0fa8988a6210e894c2af7/astroquery/gemini/tests/test_gemini.py):

> [Line 214:1](https://github.com/astropy/astroquery/blob/ade3a79574247c35b2f0fa8988a6210e894c2af7/astroquery/gemini/tests/test_gemini.py#L214): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

davner committed:
Address PEP 8 blank line.

bsipocz commented on astroquery/gemini/core.py:
```diff
@@ -433,6 +434,97 @@ def get_file(self, filename, *, download_dir='.', timeout=None):
+            if length == 0:
+                log.warn(f'URL {url} has length=0')
+
+        blocksize = astropy.utils.data.conf.download_block_size
```
you need to explicitly import this, importing `astropy` above is not enough.

bsipocz commented on astroquery/gemini/core.py:
```diff
@@ -433,6 +434,97 @@ def get_file(self, filename, *, download_dir='.', timeout=None):
         local_filepath = os.path.join(download_dir, filename)
         self._download_file(url=url, local_filepath=local_filepath, timeout=timeout)
 
+    def _download_file_content(self, url, timeout=None, auth=None, method="GET", **kwargs):
```
Make optional arguments kwarg only. Also, what are the possible keys in kwargs, explicitly list them if possible.
```suggestion
    def _download_file_content(self, url, *, timeout=None, auth=None, method="GET", **kwargs):
```

davner replied:
The kwargs could be any accepted keys in  the underlying `Session.request`. I can make a note of them.

bsipocz commented on astroquery/gemini/core.py:
```diff
@@ -433,6 +434,97 @@ def get_file(self, filename, *, download_dir='.', timeout=None):
+        self._authenticated = False
+
+    def get_file_content(self, filename, timeout=None, auth=None, method="GET", **kwargs):
+        """Wrapper around `_download_file_content`.
```
Use double backticks for everything that is not sphinx linkable.

bsipocz commented on astroquery/gemini/core.py:
```diff
@@ -433,6 +434,97 @@ def get_file(self, filename, *, download_dir='.', timeout=None):
+        str
+            The URL where the file can be downloaded.
+        """
+        return f"https://archive.gemini.edu/file/{filename}"
```
This url is hardwired at a lot of places, couldn't ve instead use the conf value, or store it as an instance attribute and use it here and elsewhere, too?

davner replied:
I agree. I have refactored the `gemini` module to use `conf` more and extend the module.

I did not want to make much changes on this first PR so I could first confirm that the documentation and coding styling matched `astroquery`.

bsipocz reviewed (changes requested):
There are a lot of test failures, those need to be fixed before this can go ahead. Also, will need a changelog entry.

(run `pytest -P gemini -R` to run the failing remote tests)


Big picture thing, that may not be known outside of a small group is that we're working on moving adding some download utilities to pyvo and plan to use that in astroquery. So it maybe useful for this PR/work, too (but given it's not yet merged, let alone released, I would not wait for it with this PR)

Besides I left a couple of smaller review comments.

davner commented:
@bsipocz Thanks for the review. I am working on addressing your comments. I have comments below:

1. I had a misunderstanding on the remote tests, as they pass with `pytest astroquery/gemini/tests -m remote_data --remote-data=any` however they fail with `pytest -P gemini -R` with repeated `AttributeError: 'dict' object has no attribute 'extract_cookies'`. I need to investigate this more. 
2. I agree that some code and constants are repeated throughout the `gemini` module. I will have this addressed in the next PR I make as it extends this module to include downloading and unpacking the Gemini Observatory Archive tar files and calibration data. This PR was more to get my feet wet with contributing to `astroquery`.
